### ISSUE TITLE
Implement 3D model of dust emission

### ISSUE THREAD

**Issue by zonca** [healpy/pysm#38]
Got in touch with Ginés Martínez-Solaeche and Jacques Delabrouille about implementing the model: "A 3-D model of polarised dust emission in the Milky Way" published in https://arxiv.org/abs/1706.04162

Would like first to complete and merge #37 that we could use as a reference.

It would be nice to know:

* is there already a Python implementation of the model?
* would a frequency-based interpolation of existing maps be a suitable implementation?
* does the model have any specific requirements? e.g. needs to load huge input datasets

**Comment by delabrou** [comment]
There is no python implementation yet.
A frequency-based interpolation of existing maps would be a suitable implementation. I produced (hundreds of) maps at frequencies ranging from 1 GHz to 1 THz last year for CMB-S4, with a small Delta_nu, and uploaded them at NERSC, but I cannot find where now. One could interpolate those. They are at nside=512 if I remember correctly.
Alternatively, we could provide the model itself, i.e. the output of the code which is 6 polarised maps of dust emissions in the 6 layers at 353 GHz, six maps of spectral indices, and six maps of temperature. Then the code would have to implement the MBB laws to compute the emission at any frequency, and add-up the six contributions.
There is no specific requirement that I foresee could cause problems.

**Comment by zonca** [comment]
Thanks @delabrou,
I would rather skip the intepolation, also because we want nside 8192.

We have this models for dust:
https://github.com/healpy/pysm/blob/master/pysm/models/dust.py

Do you think it would be feasible to modify one of the classes in `dust.py` to implement your model?
We would want inputs both at nside 512 and nside 8192.

**Comment by delabrou** [comment]
I think it might be doable to modify one of the classes. What we need is just a sum of six blackbodies, i.e. six dust blackbody components.
inputs a side 512 should be no problem. We can just give maps of amplitude (IQU), temperature and spectral index for each of the six dust components, corresponding to one realisation of the model. 8192 is challenging. For now I can do 2048 and then we upgrade? Anyway at very small scales point source polarisation will dominate...

**Comment by zonca** [comment]
if you can only do nside 2048 no problem, we can use those, PySM will ud_grade them to the requested nside.

so you could start working on a pull request to this repository which adds a class to `dust.py` (or better create another file) and then add documentation and link to paper on https://github.com/healpy/pysm/blob/master/docs/models.rst

**Comment by zonca** [comment]
@delabrou any updates on this?

**Comment by delabrou** [comment]
I got back to this recently, I found a bug in my code when Nside is different from 512. I am fixing it now and checking everything. More later.

**Comment by zonca** [comment]
> I got back to this recently, I found a bug in my code when Nside is different from 512. I am fixing it now and checking everything. More later.

@delabrou any progress?

**Comment by zonca** [comment]
from @delabrou:

> https://www.apc.univ-paris7.fr/Downloads/Planck/PSM-Data/MKD-MODEL-PySM-2048/
The components/thermaldust/ sub-directory contains maps of IQU, Temperature, and spectral index in the various layers, with straightforward names. In the skyinbands/HFI folder I have observations of the model in three significant Planck frequency bands, for reference and checks.
This specific realization of the model can then be easily implemented in PySM3, by simply summing-up the contribution of all 6 layers, scaled each with MBBs with the corresponding parameters.

There are also nside 256 versions at https://www.apc.univ-paris7.fr/Downloads/Planck/PSM-Data/MKD-MODEL-PanExGal/

**Comment by brandonshensley** [comment]
In case it is useful, [here](https://gist.github.com/brandonshensley/1dda561b245cda6497eb1b5ce2b7c267) is a super simple notebook for evaluating the MKD model at any frequency using the files @delabrou provided.

**Comment by delabrou** [comment]
Hi Brendan

Thanks, this is useful. While the general amplitude of the I857/I353 ratio fluctuates with an amplitude similar to that we see on real data, It is not distributed as that of real data, so this probably requires further investigations / refinements in the future (again, along the line of correlation of color with total amplitude)…

Jacques

**Comment by zonca** [comment]
ok, started to work on this in https://github.com/galsci/pysm/pull/87

**Comment by zonca** [comment]
yes, @brandonshensley, can you please share your nside8 maps?

**Comment by zonca** [comment]
> yes, @brandonshensley, can you please share your nside8 maps?

nevermind, I just ran your notebook

**Comment by zonca** [comment]
I fixed some issues, I still get a discrepancy of a few percent, too big to be due to rounding errors. I am worried about the unit conversions.

So in @brandonshensley 's notebook if I consider pixel 0 of the first layer:

353GHz 0.06112886862365485 MJ/sr
the mbb scaling factor is: 10.402229249394816
857 GHz 0.6358765051793955 MJ/sr

In my code:

first the input is converted to `uK_RJ` (checked this conversion and it is fine)

15.967064316226011 uK_RJ

(freq / freq_ref) ** (mbb_index[i_layer][0] - 2.0)
0.7269558731393342

blackbody_ratio(freq, freq_ref, mbb_temperature[i_layer])[0]
2.4865919382728716

(Pdb) p freq, freq_ref
(857.0, 353.0)

the output of just the first layer is `28.862746358454327 uK_RJ`

```
bandpass_unit_conversion(freq, weights, self.output_unit)
<Quantity [0.02256491] MJy / (sr uK_RJ)>


<Quantity [0.65128514] MJy / (sr uK_RJ)>
```

it is about 4% off.

**Comment by brandonshensley** [comment]
Using that temperature, I recover from WolframAlpha your mbb factor of 2.4865919382728716 to seven digits to the right of the decimal. Sticking to MJy/sr and a 353 GHz flux density of 0.06112886, I arrive at 0.651285 MJy/sr at 857 GHz, which is what you quote above. Not sure if that helps.

**Comment by zonca** [comment]
Yes, I also swapped the units with the astropy contants but got same
result. I can check more carefully later.

**Comment by brandonshensley** [comment]
Mea culpa, I swapped nu and T in my B_nu function:

`B_nu(temp, nu)/B_nu(temp,nu_ref)`

instead of

`B_nu(nu, temp)/B_nu(nu_ref, temp)`

That should fix it.
